### ISSUE TITLE
pip install fails when using pep517 // need guidance from pypa as to appropriateness of pyproject.toml file

### ISSUE THREAD

**Issue by guisen4332** [sqlalchemy/sqlalchemy#5207]
I have installed python3 python3-dev python3-pip
I tried to 'python3 -m pip install sqlalchemy'
but got this:

  Installing build dependencies ... done
  Getting requirements to build wheel ... error
  ERROR: Command errored out with exit status 1:
   command: /usr/bin/python3 /usr/share/python-wheels/pep517-0.7.0-py2.py3-none-any.whl/pep517/_in_process.py get_requires_for_build_wheel /tmp/tmptd6qlcwy
       cwd: /tmp/pip-install-216o45ia/sqlalchemy
  Complete output (1 lines):
  /usr/bin/python3: can't find '__main__' module in '/usr/share/python-wheels/pep517-0.7.0-py2.py3-none-any.whl/pep517/_in_process.py'
  ----------------------------------------
ERROR: Command errored out with exit status 1: /usr/bin/python3 /usr/share/python-wheels/pep517-0.7.0-py2.py3-none-any.whl/pep517/_in_process.py get_requires_for_build_wheel /tmp/tmptd6qlcwy Check the logs for full command output.

BTW,  'apt install python3-sqlalchemy' works.

**Comment by CaselIT** [comment]
Could you please provide your python version, pip version and which operating system are you using.
Also the logs would be useful.

**Comment by CaselIT** [comment]
I've tried using docker, and I can reproduce it.

The log with the verbose flag is
```
  Collecting wheel
    Created temporary directory: /tmp/pip-unpack-umroyu83
    Looking up "https://files.pythonhosted.org/packages/8c/23/848298cccf8e40f5bbb59009b32848a4c38f4e7f3364297ab3c3e2e2cd14/wheel-0.34.2-py2.py3-none-any.whl" in the cache
    Current age based on date: 17
    Ignoring unknown cache-control directive: immutable
    Freshness lifetime from max-age: 365000000
    The response is "fresh", returning cached response
    365000000 > 17
    Using cached wheel-0.34.2-py2.py3-none-any.whl (26 kB)
    Added wheel from https://files.pythonhosted.org/packages/8c/23/848298cccf8e40f5bbb59009b32848a4c38f4e7f3364297ab3c3e2e2cd14/wheel-0.34.2-py2.py3-none-any.whl#sha256=df277cb51e61359aba502208d680f90c0493adec6f0e848af94948778aed386e to build tracker '/tmp/pip-req-tracker-hgyx4x20'
    Removed wheel from https://files.pythonhosted.org/packages/8c/23/848298cccf8e40f5bbb59009b32848a4c38f4e7f3364297ab3c3e2e2cd14/wheel-0.34.2-py2.py3-none-any.whl#sha256=df277cb51e61359aba502208d680f90c0493adec6f0e848af94948778aed386e from build tracker '/tmp/pip-req-tracker-hgyx4x20'
  Installing collected packages: setuptools, wheel
    Created temporary directory: /tmp/pip-unpacked-wheel-ljma2ot5

    Creating /tmp/pip-build-env-0t7etson/overlay/bin
    changing mode of /tmp/pip-build-env-0t7etson/overlay/bin/easy_install to 755
    changing mode of /tmp/pip-build-env-0t7etson/overlay/bin/easy_install-3.8 to 755
    Created temporary directory: /tmp/pip-unpacked-wheel-buyul7rw

    changing mode of /tmp/pip-build-env-0t7etson/overlay/bin/wheel to 755
  Successfully installed setuptools-46.0.0 wheel-0.34.2
  Cleaning up...
  Removed build tracker: '/tmp/pip-req-tracker-hgyx4x20'
  Installing build dependencies ... done
  Running command /usr/bin/python3 /usr/share/python-wheels/pep517-0.7.0-py2.py3-none-any.whl/pep517/_in_process.py get_requires_for_build_wheel /tmp/tmpr32pf5w7
  /usr/bin/python3: can't find '__main__' module in '/usr/share/python-wheels/pep517-0.7.0-py2.py3-none-any.whl/pep517/_in_process.py'
  Getting requirements to build wheel ... error
Cleaning up...
  Removing source in /tmp/pip-install-_z4zmk8f/sqlalchemy
Removed sqlalchemy from https://files.pythonhosted.org/packages/8c/30/4134e726dd5ed13728ff814fa91fc01c447ad8700504653fe99d91fdd34b/SQLAlchemy-1.3.15.tar.gz#sha256=c4cca4aed606297afbe90d4306b49ad3a4cd36feb3f87e4bfd655c57fd9ef445 from build tracker '/tmp/pip-req-tracker-hgyx4x20'
Removed build tracker: '/tmp/pip-req-tracker-hgyx4x20'
ERROR: Command errored out with exit status 1: /usr/bin/python3 /usr/share/python-wheels/pep517-0.7.0-py2.py3-none-any.whl/pep517/_in_process.py get_requires_for_build_wheel /tmp/tmpr32pf5w7 Check the logs for full command output.
Exception information:
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/pip/_internal/cli/base_command.py", line 186, in _main
    status = self.run(options, args)
  File "/usr/lib/python3/dist-packages/pip/_internal/commands/install.py", line 357, in run
    resolver.resolve(requirement_set)
  File "/usr/lib/python3/dist-packages/pip/_internal/legacy_resolve.py", line 177, in resolve
    discovered_reqs.extend(self._resolve_one(requirement_set, req))
  File "/usr/lib/python3/dist-packages/pip/_internal/legacy_resolve.py", line 333, in _resolve_one
    abstract_dist = self._get_abstract_dist_for(req_to_install)
  File "/usr/lib/python3/dist-packages/pip/_internal/legacy_resolve.py", line 282, in _get_abstract_dist_for
    abstract_dist = self.preparer.prepare_linked_requirement(req)
  File "/usr/lib/python3/dist-packages/pip/_internal/operations/prepare.py", line 515, in prepare_linked_requirement
    abstract_dist = _get_prepared_distribution(
  File "/usr/lib/python3/dist-packages/pip/_internal/operations/prepare.py", line 95, in _get_prepared_distribution
    abstract_dist.prepare_distribution_metadata(finder, build_isolation)
  File "/usr/lib/python3/dist-packages/pip/_internal/distributions/sdist.py", line 38, in prepare_distribution_metadata
    self._setup_isolation(finder)
  File "/usr/lib/python3/dist-packages/pip/_internal/distributions/sdist.py", line 96, in _setup_isolation
    reqs = backend.get_requires_for_build_wheel()
  File "/usr/share/python-wheels/pep517-0.7.0-py2.py3-none-any.whl/pep517/wrappers.py", line 151, in get_requires_for_build_wheel
    return self._call_hook('get_requires_for_build_wheel', {
  File "/usr/share/python-wheels/pep517-0.7.0-py2.py3-none-any.whl/pep517/wrappers.py", line 245, in _call_hook
    self._subprocess_runner(
  File "/usr/lib/python3/dist-packages/pip/_internal/utils/subprocess.py", line 271, in runner
    call_subprocess(
  File "/usr/lib/python3/dist-packages/pip/_internal/utils/subprocess.py", line 242, in call_subprocess
    raise InstallationError(exc_msg)
pip._internal.exceptions.InstallationError: Command errored out with exit status 1: /usr/bin/python3 /usr/share/python-wheels/pep517-0.7.0-py2.py3-none-any.whl/pep517/_in_process.py get_requires_for_build_wheel /tmp/tmpr32pf5w7 Check the logs for full command output.
```

You can add `--no-use-pep517` to pip as a workaround for now

**Comment by guisen4332** [comment]
it works.Thanks

**Comment by CaselIT** [comment]
I'll reopen it, since it's probably best to fix this in the setup script

**Comment by zzzeek** [comment]
what is the bug?

**Comment by CaselIT** [comment]
pip install for sqlalchemy fails on docker debian:unstable. To reproduce:
```
apt update
apt install python3 python3-dev python3-pip
python3 -m pip install sqlalchemy 
````

**Comment by zzzeek** [comment]
I got that part.  but why?  what is non-compliant about our setup.py ?

**Comment by CaselIT** [comment]
I'm not sure. Maybe a new version of setup tools enables pep 517 by default?

It may just be that setup.py needs to opt-out of pep517?

**Comment by webknjaz** [comment]
It's not setuptools that enables pep517, it's pip that calls PEP517 backend.

**Comment by webknjaz** [comment]
@CaselIT is that the full log? Usually, when that breaks it has another exception from the subprocess.

**Comment by CaselIT** [comment]
that is the log running pip with the verbose output

**Comment by webknjaz** [comment]
I'm pretty sure that pip/python's broken there.

**Comment by webknjaz** [comment]
Also, the best way to fight this is to publish OS-specific wheels matrix so that building from source won't be triggered.

**Comment by CaselIT** [comment]
that was also my thought. Since it seems that it's using pep517 from `/usr/share/python-wheels/`, but I've yet to test if reinstalling pep517 with pip helps.

it may be that the python package in debian unstable is broken

**Comment by CaselIT** [comment]
I've tried reinstalling pip with `python3 -m pip install -I pip` and the sqlalchemy successfully installs so I think that the problem is in the debian version installed with apt. 

Also reading https://twitter.com/codewithanthony/status/1240690187543298048
>debian's wheel hacks are broken again (any from-source pep517 package breaks in the same way).  the presence of pyproject.toml (even though it's just for black) is triggering pip to do that.  a workaround is `pip install --no-use-pep517`

Seems to suggest that pyproject.toml may be why pip tries to use pep517. 
We could maybe exclude it from the archive on pypi.

Closing this. Thank for the help @webknjaz

**Comment by zzzeek** [comment]
we cant build wheels because SQLAlchemy includes native code and we dont have the resources to maintain native builds for all OSes.

**Comment by webknjaz** [comment]
@zzzeek well, I hear you. But in my experience, the hard part is to set up CI/CD for this. Keeping it working after that doesn't add much complexity. If you'd be open to this, I could craft a PR with scripts for this.


@CaselIT
> Seems to suggest that pyproject.toml may be why pip tries to use pep517.

Yes, this is how the latest versions of pip work.

> Since it seems that it's using pep517 from `/usr/share/python-wheels/`, but I've yet to test if reinstalling pep517 with pip helps.

The official pip is shipped with pep517 vendored. But Debian derivatives are famous for having weird modifications to the CPython distribution and python package dists, decoupling them and spreading across their fs layout. This oftentimes leads to breaking assumptions.

**Comment by zzzeek** [comment]
> @zzzeek well, I hear you. But in my experience, the hard part is to set up CI/CD for this. Keeping it working after that doesn't add much complexity. If you'd be open to this, I could craft a PR with scripts for this. Also, if you cover at least a few versions that'd be already a huge win and others'd use sdist as a fallback.

what kind of "scripts" ?  how will they run? where will they run?  where will windows builds occur ?

**Comment by zzzeek** [comment]
@rvanlaar please continue here, pyproject.toml was added because I was urged to do so. nobody was aware that this has implications for tools like "poetry" and I am not about to keep making random changes without top level guidance from pypa / people like @dstufft  as to what our best practice should be here.  When the best practice is identified with consensus, we will go for that.

**Comment by webknjaz** [comment]
> what kind of "scripts" ? how will they run? where will they run? where will windows builds occur ?

I usually do this as CI configs (Travis CI/Appveyor) with shell scripts for manylinux. The result is dists ready to be distributed via PyPI.
Normally, I set up automatic publishing to PyPI from Travis CI and AppVeyor or GitHub Actions.
Alternatively, I could add some GitHub Actions jobs and save such dists as artifacts attached to builds that can be downloaded via GitHub UI (that is if you want to do manual checks and/or upload them yourself).

**Comment by zzzeek** [comment]
we had a very long and utlimately unsuccessful attempt to get Appveyor to work some years ago .  I don't have the resources to work on this and if people don't have build tools installed, the source distribution builds pure python.   There really should not be an issue.

**Comment by webknjaz** [comment]
Honestly, nowadays I'm trying to migrate away from AppVeyor/Travis CI because of their limited resources. GitHub Actions has good support for Windows/Linux/macOS.
Another reason to have wheels is that they speed-up the installation too.

**Comment by zzzeek** [comment]
Sure wheels are great but I don't have the time / servers to build / maintain/ babysit all of that.  look what a crapshow there is just adding one file to the project for a code formatting tool.  

There should be a central service provided by or alongside pypi itself that simply builds wheels for projects on various backends and publishes them into the package page automatically.   this is not somethiugn everyone should have to be scripting on their own if they provide a package that builds with standard "python setup.py bdist_wheel".

**Comment by webknjaz** [comment]
> if they provide a package that builds with standard "python setup.py bdist_wheel"

It's not standard anymore. pip calls what's in build-system->build-backend in `pyproject.toml`. Also, PEP517 implies opt-in for using this standard by just having `pyproject.toml` present.

**Comment by zzzeek** [comment]
OK if pypi provides a wheel-building service then i will transition to whatever standard is required.

**Comment by webknjaz** [comment]
AFAIK they're slowly preparing some APIs for third-party integrations to be able to perform such builds.

**Comment by zzzeek** [comment]
per https://github.com/python-poetry/poetry/issues/2197 they seem to be able to resolve by resetting their poetry environments and per https://github.com/pypa/pip/issues/7874#issuecomment-601449946 we are compliant with best practices.

**Comment by zzzeek** [comment]
will evaluate not including pyproject.toml file as this is [documented](https://pip.pypa.io/en/stable/reference/pip/#pep-517-and-518-support) as a positive configuration step that one wishes to use pep517, and we do not have any pep517 backend built out yet.

**Comment by sqla-tester** [comment]
**Mike Bayer** has proposed a fix for this issue in the **master** branch:

**Remove pyproject.toml from distribution** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/1806

**Comment by sqla-tester** [comment]
**Mike Bayer** has proposed a fix for this issue in the **rel_1_3** branch:

**Remove pyproject.toml from distribution** https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/1807

**Comment by altendky** [comment]
setuptools is a valid build backend. No need to write your own when you already use setuptools.  And yeah, somewhere is another ticket about building wheels that I commented in but the response was so strongly negative and focused on problems rather than solutions and the setup for sqla CI so different from so many other projects that I decided not to try. But yeah, GitHub Actions with cibuildwheel would probably be pretty good in any number of other projects.

**Comment by zzzeek** [comment]
hi -

we have to write a backend because SQLAlchemy builds native code that needs to gracefully degrade to pure python if the C build does not have the right tooling available or if it otherwise fails, as is the norm on Windows machines.  There are also environment variables that control whether or not C extensions should be built, or force-built, etc.  See all the code at https://github.com/sqlalchemy/sqlalchemy/blob/master/setup.py#L189.  That all has to be ported to pep517.

**Comment by zzzeek** [comment]
also I dont know that my response to wheels is "negative".  I would love to have a whole list of wheels on our project page.   I will gladly give pypi access to individuals or organizations who wish to place wheel files for various architecutres on our pypi page.   I just dont have the resources to maintain this myself nor should I have to, just like I don't maintain Debian, Conda, Fedora binaries either.   There's supposed to be a community here.

**Comment by altendky** [comment]
Why wouldn't the setuptools backend treat your setup.py the same as it does now?

**Comment by altendky** [comment]
And yes, a community. Not somewhere where you dump the build off on someone else sounding like you are unwilling to ever touch it.  I was offering to (at least consider) setting it up, as part of the community.  Rather than looking to work together you respond as if sqla is somehow unique when it comes to wheels even though lots of other (C including) projects pull it off.  Sorry I'm being a bit whiny here but I just didn't foresee satisfying you or getting any integration into your 'community' to help with continued support.  Given that I don't even use sqla, I decided to avoid the hassle.

Anyways, maybe I can get back on topic here.  What involvement are you willing to have in the setup and maintenance of a wheel build beyond allowing PyPI access?  I'm assuming someone else had to do the grunt work for the CI, which is fine, but I also don't imagine it being integrated to whatever degree without you wanting to help on that front.

**Comment by zzzeek** [comment]
> Why wouldn't the setuptools backend treat your setup.py the same as it does now?

the goal of pep517 is to replace the need for setup.py.   I don't see the reason to interact with this pep unless I am seeking to use it fully.

**Comment by zzzeek** [comment]
> And yes, a community. Not somewhere where you dump the build off on someone else sounding like you are unwilling to ever touch it. I was offering to (at least consider) setting it up, as part of the community. Rather than looking to work together you respond as if sqla is somehow unique when it comes to wheels even though lots of other (C including) projects pull it off. Sorry I'm being a bit whiny here but I just didn't foresee satisfying you or getting any integration into your 'community' to help with continued support. Given that I don't even use sqla, I decided to avoid the hassle.

I'm really not understanding why this is so hard to explain.       I spent many, many hours / days working with someone on getting appveyor to build a windows wheel.  I had to get involved with batchfiles, powershell, all kinds of technologies to try to figure out how to get a particular workflow happening, and at the end of the day, it failed - there were silly syntactical things in the .bat / powershell files that prevented me from being able to make them work they way I wanted.  The contributor lost interest and went away, and we don't have windows wheels.       

So if today, someone went over to appveyor and built me a build system and said, "here it is, I'm done", then they are effectively "dumping" their system on *me*.   and I simply don't have the resources to debug when some change in windows means someday people are posting bug reports that they don't have this .dll or another, I have zero ability / time / interest in hunting down windows visual C stuff for people.    There are many, many people that know all about windows and building Python on windows.  I should not have to be one of those people, I have a different job.

It would simply be unwise and a disservice to the community for me to publish software that I am not capable of maintaining, surely this should be easy to understand?

> 
> Anyways, maybe I can get back on topic here. What involvement are you willing to have in the setup and maintenance of a wheel build beyond allowing PyPI access?

I maintain the source code for SQLAlchemy.  Whatever the source code needs in order to build correctly on all backends, that's my job.      When some weird issue in Python has someone having issues about MSVC or dlls or strange debian dependency issues that I have no idea about, that should be someone else.

> I'm assuming someone else had to do the grunt work for the CI, which is fine, but I also don't imagine it being integrated to whatever degree without you wanting to help on that front.

An external wheel-building system need only watch the pypi directory for new releases, and when one is present, it builds all the wheels and puts them there.  if they are failing, then they post an issue here, I fix the source code and re-release, then the wheels can come in.    if someone downloaded a wheel and there's some strange issue where it doesnt work in some kind of container or dll environment, I can CC that issue to the person who does the wheel building.

this is really not any kind of novel workflow at all.   this is exactly the workflow of distros.   I work for Red Hat.  If someone is having a problem with the Red Hat .rpm of SQLAlchemy that is only in the .rpm and not in the source distribution, Red Hat people fix that.  

I'm sorry that I'm not receptive to prescriptive advice in this area.    I'm well aware of how I could build a whole wheel-building environment.    I can pretty much build anything however I've learned over many years that just because you can build anything, you have to have some strong self-discipline that you don't go off and build *everything*.   I'm not that young anymore.

**Comment by davidism** [comment]
I understand and agree it would be preferable to have a system that built wheels for us. It is way, way harder than it should be to coordinate everything. So don't take this as suggesting you need to change, just want to throw in my experience.

In case you or someone supporting SQLAlchemy wheels in the future is interested in looking at a working example, I did put in the work to figure it out for MarkupSafe. I'm not super familiar with SQLAlchemy's speedups, but I think they're similar cases.

We use setuptools and have a fallback if the speedups can't be built. I used Azure Pipelines and cibuildwheel, so there is one config to run all the tests and build the wheels for the major platforms. I still download the artifacts then push them to PyPI, but that can be automated with a couple extra lines.

https://github.com/pallets/markupsafe/blob/32b2a46f04e7752b5aedb05f0e1b3864e16fcee3/.azure-pipelines.yml#L52-L81

**Comment by altendky** [comment]
I'm not convinced that I am 'young' either and I'm definitely not 'single'.  *sigh*

**Comment by graingert** [comment]
appveyor may not be the best choice now that GitHub actions has windows support (on Azure because Microsoft)

**Comment by zzzeek** [comment]
here's the debian bug:

https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=954256

**Comment by CaselIT** [comment]
I've also tried to setup a github action to build wheels for sqlalchemy. Current progress here: https://github.com/CaselIT/sqlalchemy/blob/action-publish/.github/workflows/create-wheels.yaml
Hopefully future version will have os specific wheels

**Comment by webknjaz** [comment]
@CaselIT FYI https://github.com/pypa/gh-action-pypi-publish. Also, I recommend you building all dists in jobs, then putting them into build artifacts and having one last job downloading all the artifacts and doing the publishing. This will avoid the situation when one job published something and others didn't.

**Comment by CaselIT** [comment]
Currently I've found using artifacts is a bit awkward at the moment:
- There seems to be no way to download all artifacts with a single action. The download artifacts v2 can do it but it currently has a bug regarding zip files. https://github.com/actions/download-artifact/issues/23#issuecomment-605435160
- I have not found a way to make them expire and if IRRC there is only 500MB of storage in free accounts.

@webknjaz I've tried that action, but it will only runs on linux.

Also I'm not sure if it would be an issue if only some jobs publish wheels. I mean, the job that have success have created working wheels so they can be published. The ones that fail hopefully can be retried. At worst there will be the source distribution.
